docs/CONFIGURATION.md
# Configuration

After installing your agents, you will need to edit a configuration file in the `etc/`directory. Pattoo will read any `.yaml` files found in this directory for configuration parameters.

For the sake of simplicity we will assume there is one file called `etc/config.yaml`.

Make sure you have configured the `main` and `pattoo-api-agentd` sections of `etc/config.yaml` file before adding any sections for `pattoo-os` related daemons.

## Custom Directory Location
You can selectively set the location of the configuration directory by using the `PATTOO_CONFIGDIR` environmental variable.

This can be useful for:

1. Automated deployments
1. Software developer code testing

By default the `etc/` directory of the repository is used for all configuration file searches.

## Mandatory Configuration Sections
```yaml
main:
    log_level: debug
    log_directory: ~/GitHub/pattoo-agents/log
    cache_directory: ~/GitHub/pattoo-agents/cache
    language: en
    polling_interval: 300

pattoo-api-agentd:
    api_ip_address: 192.168.1.100
    api_ip_bind_port: 6000
    api_uses_https: False
    api_listen_address: 0.0.0.0    

```

### Configuration Explanation

This table outlines the purpose of each configuration parameter

|Section | Config Options          | Description                    |
|--|--|--|
| `main` |||
||  `log_directory` | Path to logging directory. Make sure the username running the daemons have RW access to files there. |
||  `log_level` | Default level of logging. `debug` is best for troubleshooting. |
|| `cache_directory` | Directory of unsuccessful data posts to `pattoodb`|
|| `language` | Language  to be used in reporting statistics in JSON output. Language files can be found in the `metadata/language/agents/` directory.|
|| `polling_interval`              | Interval of data collection and posting in seconds   |
| `pattoo-api-agentd` || **Note** The `pattoo-api-agentd` section is not required for `patoo-os-spoked` configurations|
|| `api_ip_address`       | IP address of remote `pattoodb` server      |
|| `api_ip_bind_port`       | Port of remote `pattoodb` server     |
|| `api_uses_https`      | Use `https` when sending data  to remote `pattoodb` server|
|| `api_listen_address` | IP address on which the API server will listen. Setting this to `0.0.0.0` will make it listen on all IPv4 addresses. Setting to `"0::"` will make it listen on all IPv6 configured interfaces. It will not listen on IPv4 and IPv6 addresses simultaneously. You must **quote** all IPv6 addresses. The default is `0.0.0.0`|


## Agent Configuration
See the [README.md](README.md) file for details on how to configure each type of agent.

## Notes
Here are some addtional tips.

1. You can create a separate configuration file for each section. If you are doing this, make sure there is only one file per agent section. Keep the mandtatory configurations sections in a separate file for simplicity. Practice on a test system before doing this. *Start with a single file first to gain confidence.*


docs/CONFIGURATION.rst

Configuration
=============

After installing your agents, you will need to edit a configuration file in the ``etc/``\ directory. Pattoo will read any ``.yaml`` files found in this directory for configuration parameters.

For the sake of simplicity we will assume there is one file called ``etc/config.yaml``.

Make sure you have configured the ``main`` and ``pattoo-api-agentd`` sections of ``etc/config.yaml`` file before adding any sections for ``pattoo-os`` related daemons.

Custom Directory Location
-------------------------

You can selectively set the location of the configuration directory by using the ``PATTOO_CONFIGDIR`` environmental variable.

This can be useful for:


#. Automated deployments
#. Software developer code testing

By default the ``etc/`` directory of the repository is used for all configuration file searches.

Mandatory Configuration Sections
--------------------------------

.. code-block:: yaml

   main:
       log_level: debug
       log_directory: ~/GitHub/pattoo-agents/log
       cache_directory: ~/GitHub/pattoo-agents/cache
       daemon_directory: ~/GitHub/pattoo-agents/daemon
       language: en
       polling_interval: 300

   pattoo-api-agentd:
       api_ip_address: 192.168.1.100
       api_ip_bind_port: 6000
       api_uses_https: False
       api_listen_address: 0.0.0.0

Configuration Explanation
^^^^^^^^^^^^^^^^^^^^^^^^^

This table outlines the purpose of each configuration parameter

.. list-table::
   :header-rows: 1

   * - Section
     - Config Options
     - Description
   * - ``main``
     - 
     - 
   * - 
     - ``log_directory``
     - Path to logging directory. Make sure the username running the daemons have RW access to files there.
   * - 
     - ``log_level``
     - Default level of logging. ``debug`` is best for troubleshooting.
   * - 
     - ``cache_directory``
     - Directory of unsuccessful data posts to ``pattoodb``
   * - 
     - ``daemon_directory``
     - Directory used to store daemon related data that needs to be maintained between reboots
   * - 
     - ``language``
     - Language  to be used in reporting statistics in JSON output. Language files can be found in the ``metadata/language/agents/`` directory.
   * - 
     - ``polling_interval``
     - Interval of data collection and posting in seconds
   * - ``pattoo-api-agentd``
     - 
     - **Note** The ``pattoo-api-agentd`` section is not required for ``patoo-os-spoked`` configurations
   * - 
     - ``api_ip_address``
     - IP address of remote ``pattoodb`` server
   * - 
     - ``api_ip_bind_port``
     - Port of remote ``pattoodb`` server
   * - 
     - ``api_uses_https``
     - Use ``https`` when sending data  to remote ``pattoodb`` server
   * - 
     - ``api_listen_address``
     - IP address on which the API server will listen. Setting this to ``0.0.0.0`` will make it listen on all IPv4 addresses. Setting to ``"0::"`` will make it listen on all IPv6 configured interfaces. It will not listen on IPv4 and IPv6 addresses simultaneously. You must **quote** all IPv6 addresses. The default is ``0.0.0.0``


Agent Configuration
-------------------

See the `README.rst <README.rst>`_ file for details on how to configure each type of agent.

Notes
-----

Here are some addtional tips.


#. You can create a separate configuration file for each section. If you are doing this, make sure there is only one file per agent section. Keep the mandtatory configurations sections in a separate file for simplicity. Practice on a test system before doing this. *Start with a single file first to gain confidence.*


pattoo_shared/configuration.py
#!/usr/bin/env python3
"""Pattoo classes that manage various configurations."""

# Standard imports
import os

# Import project libraries
from pattoo_shared import files
from pattoo_shared import log
from pattoo_shared.constants import (
    PATTOO_API_AGENT_PREFIX, PATTOO_API_AGENT_EXECUTABLE)


class Config(object):
    """Class gathers all configuration information.

    Only processes the following YAML keys in the configuration file:

        main:
        remote_api:

    """

    def __init__(self):
        """Initialize the class.

        Args:
            None

        Returns:
            None

        """
        # Get the configuration directory
        # Expand linux ~ notation for home directories if provided.
        _config_directory = log.check_environment()
        config_directory = os.path.expanduser(_config_directory)
        self._configuration = files.read_yaml_files(config_directory)

    def api_listen_address(self):
        """Get api_listen_address.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = PATTOO_API_AGENT_EXECUTABLE
        sub_key = 'api_listen_address'
        result = search(key, sub_key, self._configuration, die=False)

        # Default to 0.0.0.0
        if result is None:
            result = '0.0.0.0'
        return result

    def polling_interval(self):
        """Get interval.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        key = 'main'
        sub_key = 'polling_interval'
        intermediate = search(key, sub_key, self._configuration, die=False)

        # Default to 300
        if intermediate is None:
            result = 300
        else:
            result = int(intermediate)
        return result

    def api_ip_address(self):
        """Get api_ip_address.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = PATTOO_API_AGENT_EXECUTABLE
        sub_key = 'api_ip_address'

        # Get result
        result = search(key, sub_key, self._configuration, die=False)
        if result is None:
            result = 'localhost'
        return result

    def api_ip_bind_port(self):
        """Get api_ip_bind_port.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = PATTOO_API_AGENT_EXECUTABLE
        sub_key = 'api_ip_bind_port'

        # Get result
        intermediate = search(key, sub_key, self._configuration, die=False)
        if intermediate is None:
            result = 6000
        else:
            result = int(intermediate)
        return result

    def api_uses_https(self):
        """Get api_uses_https.

        Args:
            None

        Returns:
            result: result

        """
        # Initialize key variables
        key = PATTOO_API_AGENT_EXECUTABLE
        sub_key = 'api_uses_https'

        # Get result
        result = search(key, sub_key, self._configuration, die=False)
        if result is None:
            result = False
        return result

    def api_uri(self):
        """Get api_uri.

        Args:
            None

        Returns:
            result: result

        """
        # Return
        result = '{}/receive'.format(PATTOO_API_AGENT_PREFIX)
        return result

    def api_server_url(self, agent_id):
        """Get pattoo server's remote URL.

        Args:
            agent_id: Agent ID

        Returns:
            result: URL.

        """
        # Construct URL for server
        if self.api_uses_https() is True:
            prefix = 'https://'
        else:
            prefix = 'http://'

        # Return
        result = (
            '{}{}:{}{}/{}'.format(
                prefix, self.api_ip_address(),
                self.api_ip_bind_port(), self.api_uri(), agent_id))
        return result

    def log_directory(self):
        """Get log_directory.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        sub_key = 'log_directory'
        result = None
        key = 'main'

        # Get new result
        _result = search(key, sub_key, self._configuration)

        # Expand linux ~ notation for home directories if provided.
        result = os.path.expanduser(_result)

        # Check if value exists. We cannot use log2die_safe as it does not
        # require a log directory location to work properly
        if os.path.isdir(result) is False:
            log_message = (
                'log_directory: "{}" '
                'in configuration doesn\'t exist!'.format(result))
            log.log2die_safe(1003, log_message)

        # Return
        return result

    def log_file(self):
        """Get log_file.

        Args:
            None

        Returns:
            result: result

        """
        _log_directory = self.log_directory()
        result = '{}{}pattoo.log'.format(_log_directory, os.sep)
        return result

    def log_file_api(self):
        """Get log_file_api.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        _log_directory = self.log_directory()
        result = '{}{}pattoo-api.log'.format(_log_directory, os.sep)
        return result

    def log_level(self):
        """Get log_level.

        Args:
            None

        Returns:
            result: result

        """
        # Get result
        sub_key = 'log_level'
        key = 'main'
        result = None

        # Return
        intermediate = search(key, sub_key, self._configuration, die=False)
        if intermediate is None:
            result = 'debug'
        else:
            result = '{}'.format(intermediate).lower()
        return result

    def cache_directory(self):
        """Determine the cache_directory.

        Args:
            None

        Returns:
            value: configured cache_directory

        """
        # Initialize key variables
        key = 'main'
        sub_key = 'cache_directory'

        # Get result
        _value = search(key, sub_key, self._configuration)

        # Expand linux ~ notation for home directories if provided.
        value = os.path.expanduser(_value)

        # Create directory if it doesn't exist
        files.mkdir(value)

        # Return
        return value

    def daemon_directory(self):
        """Determine the daemon_directory.

        Args:
            None

        Returns:
            value: configured daemon_directory

        """
        # Initialize key variables
        key = 'main'
        sub_key = 'daemon_directory'

        # Get result
        _value = search(key, sub_key, self._configuration)

        # Expand linux ~ notation for home directories if provided.
        value = os.path.expanduser(_value)

        # Create directory if it doesn't exist
        files.mkdir(value)

        # Return
        return value

    def agent_cache_directory(self, agent_program):
        """Get agent_cache_directory.

        Args:
            agent_program: Name of agent

        Returns:
            result: result

        """
        # Get result
        result = '{}/{}'.format(self.cache_directory(), agent_program)

        # Create directory if it doesn't exist
        files.mkdir(result)

        # Return
        return result


def search(key, sub_key, config_dict, die=True):
    """Get config parameter from YAML.

    Args:
        key: Primary key
        sub_key: Secondary key
        config_dict: Dictionary to explore
        die: Die if true and the result encountered is None

    Returns:
        result: result

    """
    # Get result
    result = None

    # Verify config_dict is indeed a dict.
    # Die safely as log_directory is not defined
    if isinstance(config_dict, dict) is False:
        log.log2die_safe(1021, 'Invalid configuration file. YAML not found')

    # Get new result
    if key in config_dict:
        # Make sure we don't have a None value
        if config_dict[key] is None:
            log_message = (
                '{}: value in configuration is blank. Please fix'.format(key))
            log.log2die_safe(1004, log_message)

        # Get value we need
        if sub_key in config_dict[key]:
            result = config_dict[key][sub_key]

    # Error if not configured
    if result is None and die is True:
        log_message = (
            '{}:{} not defined in configuration'.format(key, sub_key))
        log.log2die_safe(1016, log_message)

    # Return
    return result
